### ISSUE TITLE
prometheus-config-reloader needs to retry

### ISSUE THREAD

**Issue by xgerman** [coreos/prometheus-operator#437]
**What did you do?**
I started a cluster fresh with prometheus operator. Then I deployed a Prometheus object like in the example.

**What did you expect to see?**
Prometheus to become active.

**What did you see instead? Under which circumstances?**
Prometheus doesn't seem to be active. I am seeing this in the logs:
```
logs prometheus-prometheus-0  prometheus-config-reloader
ts=2017-06-21T19:39:05Z caller=main.go:227 component=volume-watcher msg=Starting...
ts=2017-06-21T19:39:05Z caller=main.go:202 component=volume-watcher msg="Updating rule files..."
ts=2017-06-21T19:39:05Z caller=main.go:207 component=volume-watcher msg="Rule files updated."
ts=2017-06-21T19:39:05Z caller=main.go:210 component=volume-watcher msg="Reloading Prometheus..."
ts=2017-06-21T19:39:06Z caller=main.go:213 component=volume-watcher msg="Reloading Prometheus failed." err="Received response code %!s(int=404), expected 200"

logs prometheus-prometheus-0  prometheus
time="2017-06-21T19:39:04Z" level=info msg="Starting prometheus (version=1.6.1, branch=master, revision=4666df502c0e239ed4aa1d80abbbfb54f61b23c3)" source="main.go:88"
time="2017-06-21T19:39:04Z" level=info msg="Build context (go=go1.8.1, user=root@7e45fa0366a7, date=20170419-14:32:22)" source="main.go:89"
time="2017-06-21T19:39:04Z" level=info msg="Loading configuration file /etc/prometheus/config/prometheus.yaml" source="main.go:251"
time="2017-06-21T19:39:04Z" level=info msg="Loading series map and head chunks..." source="storage.go:421"
time="2017-06-21T19:39:04Z" level=info msg="0 series loaded." source="storage.go:432"
time="2017-06-21T19:39:04Z" level=info msg="Starting target manager..." source="targetmanager.go:61"
time="2017-06-21T19:39:04Z" level=info msg="Listening on :9090" source="web.go:259"
```
From my limited understanding the prometheus object is not getting the configs it needs because the config-reloader is not retrying (?)

**Environment**

* Kubernetes version information:
```
kubectl version
Client Version: version.Info{Major:"1", Minor:"6", GitVersion:"v1.6.3", GitCommit:"0480917b552be33e2dba47386e51decb1a211df6", GitTreeState:"clean", BuildDate:"2017-05-10T23:29:08Z", GoVersion:"go1.8.1", Compiler:"gc", Platform:"darwin/amd64"}
Server Version: version.Info{Major:"1", Minor:"6", GitVersion:"v1.6.4+coreos.0", GitCommit:"8996efde382d88f0baef1f015ae801488fcad8c4", GitTreeState:"clean", BuildDate:"2017-05-19T21:11:20Z", GoVersion:"go1.7.5", Compiler:"gc", Platform:"linux/amd64"}
```

* Kubernetes cluster kind: 

	tectonic-installer

* Manifests:

```
# Pulled from https://github.com/coreos/prometheus-operator/blob/master/example/user-guides/getting-started/prometheus-example.yaml
apiVersion: monitoring.coreos.com/v1alpha1
kind: Prometheus
metadata:
  name: prometheus
  namespace: test
spec:
  replicas: 3
  externalUrl: http://example.com
  matchLabels:
      monitor: test-prometheus
  version: v1.7.1
  resources:
    requests:
      memory: 400Mi
```

* Prometheus Operator Logs:

```
logs prometheus-operator-2780627485-j9dtq
ts=2017-06-21T19:22:10Z caller=operator.go:136 component=alertmanageroperator msg="connection established" cluster-version=v1.6.4+coreos.0
ts=2017-06-21T19:22:10Z caller=operator.go:220 component=prometheusoperator msg="connection established" cluster-version=v1.6.4+coreos.0
ts=2017-06-21T19:22:10Z caller=operator.go:536 component=alertmanageroperator msg="TPR created" tpr=alertmanager.monitoring.coreos.com
ts=2017-06-21T19:22:10Z caller=operator.go:1034 component=prometheusoperator msg="TPR created" tpr=service-monitor.monitoring.coreos.com
ts=2017-06-21T19:22:10Z caller=operator.go:1034 component=prometheusoperator msg="TPR created" tpr=prometheus.monitoring.coreos.com
ts=2017-06-21T19:22:16Z caller=operator.go:150 component=alertmanageroperator msg="TPR API endpoints ready"
ts=2017-06-21T19:22:20Z caller=operator.go:234 component=prometheusoperator msg="TPR API endpoints ready"
ts=2017-06-21T19:22:39Z caller=operator.go:271 component=prometheusoperator msg="Prometheus added" key=test-system/prometheus
ts=2017-06-21T19:22:39Z caller=operator.go:619 component=prometheusoperator msg="sync prometheus" key=test-system/prometheus
ts=2017-06-21T19:22:40Z caller=operator.go:619 component=prometheusoperator msg="sync prometheus" key=test-system/prometheus
ts=2017-06-21T19:22:40Z caller=operator.go:584 component=prometheusoperator msg="update handler" old=3143 cur=3144
ts=2017-06-21T19:22:40Z caller=operator.go:321 component=alertmanageroperator msg="update handler" old=3143 cur=3144
ts=2017-06-21T19:22:40Z caller=operator.go:584 component=prometheusoperator msg="update handler" old=3144 cur=3150
ts=2017-06-21T19:22:40Z caller=operator.go:321 component=alertmanageroperator msg="update handler" old=3144 cur=3150
E0621 19:22:40.242713       1 operator.go:526] Sync "test-system/prometheus" failed: syncing version failed: scaling in progress, 3 expected replicas, 1 found replicas
ts=2017-06-21T19:22:40Z caller=operator.go:619 component=prometheusoperator msg="sync prometheus" key=test-system/prometheus
ts=2017-06-21T19:22:40Z caller=operator.go:584 component=prometheusoperator msg="update handler" old=3150 cur=3153
ts=2017-06-21T19:22:40Z caller=operator.go:321 component=alertmanageroperator msg="update handler" old=3150 cur=3153
E0621 19:22:40.320547       1 operator.go:526] Sync "test-system/prometheus" failed: syncing version failed: scaling in progress, 3 expected replicas, 1 found replicas
```

**Comment by jamiehannaford** [comment]
@xgerman What does the manifest look like for `prometheus-prometheus-0`? You can check by running `kubectl -n rackspace-system get pod prometheus-prometheus-0 -o yaml`.

I suspect you're not including [this parameter](https://github.com/coreos/prometheus-operator/blob/fd51262d9abd71ec7bfb6014348dc8e00174286b/contrib/prometheus-config-reloader/examples/prometheus-config-reloader.yaml#L37). That URL needs to return a 200.

**Comment by brancz** [comment]
From what I'm seeing all of the above might be ok. The reloader might detect file changes before the Prometheus instance was up and running. Which is intended right now, but might change soon-ish ( #436 ).

A couple of outputs would be helpful here:

* `kubectl -n rackspace-system get statefulset prometheus-prometheus -oyaml`
* `kubectl -n rackspace-system get pod prometheus-prometheus-0`

@jamiehannaford the statefulset is managed by the Prometheus Operator, so any changes would be overridden. The statefulset is basically generated from the Prometheus object.

**Comment by xgerman** [comment]
```
kubectl -n rackspace-system get statefulset prometheus-prometheus -oyaml
apiVersion: apps/v1beta1
kind: StatefulSet
metadata:
  annotations:
    kubectl.kubernetes.io/last-applied-configuration: |
      {"apiVersion":"monitoring.coreos.com/v1alpha1","kind":"Prometheus","metadata":{"annotations":{},"name":"prometheus","namespace":"rackspace-system"},"spec":{"externalUrl":"http://geichberger-1-worker-0.pug.systems:30080/monitor","matchLabels":{"monitor":"rackspace-prometheus"},"replicas":3,"resources":{"requests":{"memory":"400Mi"}},"version":"v1.7.1"}}
  creationTimestamp: 2017-06-21T21:57:16Z
  generation: 1
  labels:
    app: prometheus
    prometheus: prometheus
  name: prometheus-prometheus
  namespace: rackspace-system
  resourceVersion: "124404"
  selfLink: /apis/apps/v1beta1/namespaces/rackspace-system/statefulsets/prometheus-prometheus
  uid: 974fb5dd-56cc-11e7-8c62-fa163ef812c6
spec:
  replicas: 3
  selector:
    matchLabels:
      app: prometheus
      prometheus: prometheus
  serviceName: prometheus-operated
  template:
    metadata:
      creationTimestamp: null
      labels:
        app: prometheus
        prometheus: prometheus
    spec:
      containers:
      - args:
        - -storage.local.retention=24h
        - -storage.local.num-fingerprint-mutexes=4096
        - -storage.local.path=/var/prometheus/data
        - -storage.local.chunk-encoding-version=2
        - -config.file=/etc/prometheus/config/prometheus.yaml
        - -storage.local.target-heap-size=279620266
        - -web.external-url=http://geichberger-1-worker-0.pug.systems:30080/monitor
        - -web.route-prefix=
        image: quay.io/prometheus/prometheus:v1.7.1
        imagePullPolicy: IfNotPresent
        livenessProbe:
          failureThreshold: 10
          httpGet:
            path: /status
            port: web
            scheme: HTTP
          initialDelaySeconds: 300
          periodSeconds: 5
          successThreshold: 1
          timeoutSeconds: 3
        name: prometheus
        ports:
        - containerPort: 9090
          name: web
          protocol: TCP
        readinessProbe:
          failureThreshold: 6
          httpGet:
            path: /status
            port: web
            scheme: HTTP
          periodSeconds: 5
          successThreshold: 1
          timeoutSeconds: 3
        resources:
          requests:
            memory: 400Mi
        terminationMessagePath: /dev/termination-log
        terminationMessagePolicy: File
        volumeMounts:
        - mountPath: /etc/prometheus/config
          name: config
          readOnly: true
        - mountPath: /etc/prometheus/rules
          name: rules
          readOnly: true
        - mountPath: /var/prometheus/data
          name: prometheus-prometheus-db
      - args:
        - -reload-url=http://localhost:9090/-/reload
        - -config-volume-dir=/etc/prometheus/config
        - -rule-volume-dir=/etc/prometheus/rules
        image: quay.io/coreos/prometheus-config-reloader:v0.0.1
        imagePullPolicy: IfNotPresent
        name: prometheus-config-reloader
        resources:
          limits:
            cpu: 5m
            memory: 10Mi
        terminationMessagePath: /dev/termination-log
        terminationMessagePolicy: File
        volumeMounts:
        - mountPath: /etc/prometheus/config
          name: config
          readOnly: true
        - mountPath: /etc/prometheus/rules
          name: rules
      dnsPolicy: ClusterFirst
      restartPolicy: Always
      schedulerName: default-scheduler
      securityContext: {}
      terminationGracePeriodSeconds: 600
      volumes:
      - name: config
        secret:
          defaultMode: 420
          secretName: prometheus-prometheus
      - emptyDir: {}
        name: rules
      - emptyDir: {}
        name: prometheus-prometheus-db
status:
  observedGeneration: 1
  replicas: 0

kubectl -n rackspace-system get pod prometheus-prometheus-0
NAME                      READY     STATUS             RESTARTS   AGE
prometheus-prometheus-0   1/2       CrashLoopBackOff   147        13h
```

**Comment by brancz** [comment]
Yes. This is a bug that was present in the Prometheus Operator v0.10.0 and v0.10.1, we released v0.10.2 yesterday, that fixes it. It's because the "-web.route-prefix" argument is not being set properly. A quick fix to make it work with your manifest and versions is to simply add the "routePrefix" field to your Prometheus manifest, and set it to "/".

If you upgrade to v0.10.2, it should work without setting this, as that is what it should default to.

**Comment by brancz** [comment]
Note that you may need to manually delete the pods after making the changes, as the StatefulSet controller does not reconcile the Pods automatically yet.
